daemon/daemon.go

```diff
@@ -214,65 +215,50 @@ func (d *Daemon) RemoveNetworkPolicy(e *endpoint.Endpoint) {
-func (d *Daemon) QueueEndpointBuild(req *endpoint.Request) {
-	go func(req *endpoint.Request) {
-		d.uniqueIDMU.Lock()
-		// We are skipping new requests, but only if the endpoint has not
-		// started its build process, since the endpoint is already in queue.
-		if isBuilding, exists := d.uniqueID[req.ID]; !isBuilding && exists {
-			req.MyTurn <- false
-		} else {
-			// We mark the request "not building" state and send it to
-			// the building queue.
-			d.uniqueID[req.ID] = false
-			d.buildEndpointChan <- req
-		}
+func (d *Daemon) QueueEndpointBuild(epID uint64) func() {
+	d.uniqueIDMU.Lock()
+	// We are skipping new requests, but only if the endpoint has not
+	// started its build process, since the endpoint is already in queue.
+	if _, queued := d.uniqueID[epID]; queued {
 		d.uniqueIDMU.Unlock()
-	}(req)
-}
+		return nil
+	}
+	// We mark the request as queueing
+	ctx, cancel := context.WithCancel(context.Background())
+	d.uniqueID[epID] = cancel
+	d.uniqueIDMU.Unlock()
 
-// RemoveFromEndpointQueue removes the endpoint from the queue.
-func (d *Daemon) RemoveFromEndpointQueue(epID uint64) {
+	// Acquire build permit
+	err := d.buildEndpointSem.Acquire(ctx, 1)
+
+	// Not queueing any more
 	d.uniqueIDMU.Lock()
 	delete(d.uniqueID, epID)
 	d.uniqueIDMU.Unlock()
+
+	if err != nil || ctx.Err() == context.Canceled {
+		return nil
+	}
+
+	// Give build permission.
+	// Caller calls the returned function when the heavy lifting of the build is done.
+	done := false
+	return func() {
+		if !done {
+			d.buildEndpointSem.Release(1)
+			done = true
+		}
+	}
 }
 
-// StartEndpointBuilders creates `nRoutines` go routines that listen on the
-// `d.buildEndpointChan` for new endpoints.
-func (d *Daemon) StartEndpointBuilders(nRoutines int) {
-	log.WithField("count", nRoutines).Debug("Creating worker threads")
-	for w := 0; w < nRoutines; w++ {
-		go func() {
-			for e := range d.buildEndpointChan {
-				d.uniqueIDMU.Lock()
-				if _, ok := d.uniqueID[e.ID]; !ok {
-					// If the request is not present in the uniqueID,
-					// it means the request was deleted from the queue
-					// so we deny the request's turn.
-					e.MyTurn <- false
-					d.uniqueIDMU.Unlock()
-					continue
-				}
-				// Set the endpoint to "building" state
-				d.uniqueID[e.ID] = true
-				e.MyTurn <- true
-				d.uniqueIDMU.Unlock()
-				// Wait for the endpoint to build
-				<-e.Done
-				d.uniqueIDMU.Lock()
-				// In a case where the same endpoint enters the
-				// building queue, while it was still being build,
-				// it will be marked as `false`/"not building",
-				// thus, we only delete the endpoint from the
-				// queue only if it is marked as isBuilding.
-				if isBuilding := d.uniqueID[e.ID]; isBuilding {
-					delete(d.uniqueID, e.ID)
-				}
-				d.uniqueIDMU.Unlock()
-			}
-		}()
+// RemoveFromEndpointQueue removes the endpoint from the queue.
+func (d *Daemon) RemoveFromEndpointQueue(epID uint64) {
```

Comment:
Oh, perhaps this needs to be called when an endpoint is removed from `endpointmanager`. A queued build probably shouldn't proceed in that case (but this might be handled in some other way... I don't think it's handled in the older code here)

Reply:
This is called from `endpoint.LeaveLocked()` and will cancel waiting for a build permit. This call will not cancel a build already in progress, but no change in this regard.